[로또] 베디 미션 제출합니다.

dpudpu opened the pull request:
감사합니다. CU님 👍

dpudpu committed:
docs: Readme.md 요구사항 작성

dpudpu committed:
feat: LottoNo.class 로또번호를 포장해주는 VO 구현

dpudpu committed:
feat: LottoNo.class 로또번호를 포장해주는 VO 구현

dpudpu committed:
feat: Money.class 로또 구입 금액을 포장해주는 VO 구현

dpudpu committed:
feat: LottoNoCollection.class LottoNo 들을 관리해주는 클래스 구현

dpudpu committed:
feat: Lotto.class 구현

- 중복 확인
- 번호 순서 오름차순 정렬

dpudpu committed:
feat: Rank.class 구현

dpudpu committed:
refactor: LottoTest.class 테스트코드 수정

dpudpu committed:
feat: LottoNosManualGenerator.class 수동 번호 생성 구현

dpudpu committed:
fix: LottoNosManualGenerator.class 상속 추가 및 반환타입 변경

dpudpu committed:
feat: LottoNosAutoGenerator.class 로또번호 자동 생성 구현

dpudpu committed:
feat: WinningLottoNo.class 당첨번호 구현

dpudpu committed:
feat: WinPrize.class 우승상금 구현

dpudpu committed:
refactor: Money.class money long으로 타입 변경

dpudpu committed:
refactor: Money.getCountOfPurchase() 네이밍 변경

dpudpu committed:
feat: InputView.class OutputView.class 구현

dpudpu committed:
feat: ConsoleUILottoApplication.class 구현

dpudpu committed:
refactor: Money.class 예외처리 추가

dpudpu committed:
feat: 수동 개수 입력이 구입금액 초과 예외처리 추가

dpudpu committed:
refactor: Money.getCountOfPurchase() 리턴 타입 int로 변경

dpudpu committed:
refactor: 수동 입력 한번에 입력되게 수정

dpudpu committed:
feat: LottosGenerator.class 구현

- 수동개수와 총 구매개수를 비교해서 유효성 검사
- 수동개수, 오토개수만큼 각각 생성
- 웹 확장을 생각해서 Controller에서 최대한 로직 없애려고 분리했으나 생성자에 InputView를 넣어야해서 살짝 고민.
- InputView를 넣어준 이유는 웹에서는 수동 입력이 다르기 때문

dpudpu committed:
refactor: CountOfManual.class 수동 횟수 원시값 포장

dpudpu committed:
refactor: LottosGenerator.class inputView 삽입 제거 후 List<String> 넣는 방식으로 변경

dpudpu committed:
test: LottoNosAutoGenratorTest.class 제대로 리턴하는지 테스트 추가

dpudpu committed:
refactor: ConsoleUILottoApplication 일부 메소드 이름 변경

dpudpu committed:
feat: LottoService.class 구현

domain 외에도 콘솔UI와 웹UI 에서 겹칠만한 코드 Service클래스로 추출

dpudpu committed:
refactor: 매직넘버 상수 추출

dpudpu committed:
fix: CountOfManual.class 예외처리 수정

dpudpu committed:
fix: 총수익률 에러 수정

woowahanCU commented on src/test/java/lotto/domain/WinPrizeTest.java:
```diff
@@ -0,0 +1,20 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class WinPrizeTest {
```
테스트가 실패합니다.

dpudpu replied:
수정했습니다 감사합니다

woowahanCU commented on src/main/java/lotto/domain/WinPrize.java:
```diff
@@ -0,0 +1,40 @@
+        for (Rank rank : ranks.keySet()) {
+            totalPrize += rank.getPrize() * ranks.get(rank);
+        }
+        System.out.println(totalPrize+"원");
```
공백 컨벤션 위반

woowahanCU replied:
도메인 모델에 출력하는 부분이 있네요. 
그리고 뷰모델에서 해당 메소드를 호출하는데요.
web ui에서는 동일 로직을 뷰모델에서 구현해야 하지 않을까요

dpudpu replied:
반영했습니다 감사합니다 ㅎㅎ

woowahanCU commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+    private final Set<LottoNo> lottoNos;
+
+    private Lotto(final List<LottoNo> lottoNos) {
+        this.lottoNos = new TreeSet<>(lottoNos);
```
👍

woowahanCU commented on src/main/java/lotto/domain/LottoNo.java:
```diff
@@ -0,0 +1,62 @@
+        for (int i = MIN_NUMBER; i <= MAX_NUMBER; i++) {
+            map.put(i, new LottoNo(i));
+        }
+    }
```
LottoNo의 Map에 별도로 넣는 이유가 있을까요?
LottoNoCollection의 List를 활용해도 되지 않을까요

dpudpu replied:
일단 List가 아닌 Map으로 한 이유는 로또가 1~45지만 만약에 
`나중에 음수가 나올수도 있지 않을까?` 라는 생각으로 Map으로 했습니다. 
그런데 좀 고민이었던 부분은 LottoNoCollection에서 또 List로 만들어서 담는게 낭비가 아닐까라는 생각은 들었는데 
그냥 LottoNo의 Map을 지우고 LottoNoCollection의 List만 남겨서 여기서 캐싱하는게 좋을까요?

woowahanCU replied:
아, Map이냐 List냐 때문에 Comment를 남긴건 아니었구요.
중복로직이어서 변경해보니 기능도 정상동작하는 거 같더라구요. ㅎㅎ;

dpudpu replied:
아~ 그렇죠.. 중복이라서 좀 신경쓰이긴 했거든요. 
피드백 감사합니다

woowahanCU commented on src/main/java/lotto/domain/LottoNoCollection.java:
```diff
@@ -0,0 +1,18 @@
+        for (int i = LottoNo.MIN_NUMBER; i <= LottoNo.MAX_NUMBER; i++) {
+            lottoNos.add(LottoNo.from(i));
+        }
+    }
```
👍

woowahanCU commented on src/main/java/lotto/domain/generator/LottoNosAutoGenerator.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.List;
+
+public class LottoNosAutoGenerator implements LottoNosGenerator {
```
👍

woowahanCU commented on src/main/java/lotto/domain/generator/LottosGenerator.java:
```diff
@@ -0,0 +1,52 @@
+
+    private void addAutoLotto(final List<Lotto> lottos) {
+        LottoNosGenerator lottoNosGenerator = new LottoNosAutoGenerator();
+        int countOfAuto = countOfPurchase - lottos.size();
```
사소한 부분이지만, 메소드가 최대한 한가지 기능만 하도록 구현해보세요.

dpudpu replied:
오.. 그러네요 
이건 미처 생각 못했는데 반영했습니다 감사합니다.

```java
private void addAutoLotto(final List<Lotto> lottos, int countOfAuto) {
  LottoNosGenerator lottoNosGenerator = new LottoNosAutoGenerator();
  for (int i = 0; i < countOfAuto; i++) {
    List<LottoNo> lottoNos = lottoNosGenerator.generate();
    lottos.add(Lotto.of(lottoNos));
  }
}

woowahanCU commented on src/test/java/lotto/domain/generator/LottosGeneratorTest.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Test
+    void 유효성_에러_확인() {
```
어떤 유효성 에러인지 테스트코드 이름에서 드러나야 합니다.

dpudpu replied:
그러네요. 저번에는 이렇게 해도 알아 보겠지 했는데 
오늘 다시 보니 이게 뭐지? 하고 코드를 읽어보게 되네요 감사합니다

woowahanCU reviewed (changes requested):
안녕하세요. 리뷰어 CU 입니다!
전반적으로 깔끔하게 잘 구현하셨네요. 💯 
몇가지 피드백 남겼으니, 확인해보시구요. 궁금하신 부분 있으시면 DM 남겨주세요.
다음 Step으로 진행하셔도 좋습니다. 😀

dpudpu committed:
Test RankTest 테스트케이스 추가

dpudpu committed:
Test WinPrizeTest 테스트 케이스 수정

dpudpu committed:
fix: WinPrize.class sout 제거

dpudpu committed:
refactor: LottoGeneratorTest 테스트명 변경

dpudpu committed:
refactor: LottosGenerator.addAutoLotto() 한 가지 일만 하도록 수정

- 기존의 addAutoLotto()는 countOfAuto를 직접 구했는데 이렇게 할 경우 두 가지 일을 하게 된다.
- 그래서 countOfAuto는 외부에서 구해서 addAutoLotto()의 파라미터로 넘겨주도록 변경

dpudpu committed:
refactor: LottoNo 캐싱을 LottoNoCollection.class가 담당하도록 변경

dpudpu reviewed (commented):
안녕하세요. CU님
피드백 다 반영했습니다. 미처 생각 못했던 부분들도 알게 되고 
꼼꼼하게 해야겠다는 생각을 한번 더 하게 됐네요. 
리뷰해주셔서 감사합니다.

woowahanCU commented on src/test/java/lotto/domain/RankTest.java:
```diff
@@ -8,7 +8,12 @@
 public class RankTest {
     @Test
     void SECOND_THIRD_등수_확인() {
```
하나의 단위테스트에 assertion을 다중으로 넣는 것이 옳을지에 대해서도 한번 생각해봐요
https://softwareengineering.stackexchange.com/questions/7823/is-it-ok-to-have-multiple-asserts-in-a-single-unit-test

dpudpu replied:
너무 생각없이 했네요
반영했습니다 감사합니다

dpudpu committed:
refactor: RankTest 테스트 단위 분리

woowahanCU reviewed (approved):
아, 이번 미션은 머지 후에 웹 UI 적용으로 넘어가는 것이었군요 ㅋㅋ
피드백 반영 감사합니다. 💯  
Merge할게요~
